src/libxively/event_dispatcher/xi_time_event.c

```diff
@@ -0,0 +1,426 @@
+    const xi_vector_elem_t* elem = xi_insert_time_event( vector, time_event );
+
+    /* if there is a problem with the memory go to err_handling */
+    XI_CHECK_MEMORY( elem, out_state );
+
+    /* extract the element */
+    xi_time_event_t* added_time_event = ( xi_time_event_t* )elem->selector_t.ptr_value;
+
+    /* sanity checks */
+    assert( added_time_event == time_event );
+
+    if ( NULL != ret_time_event_handle )
+    {
+        /* update the return value with the pointer to the position in the vector */
+        ret_time_event_handle->position = &added_time_event->position;
+
+        /* set the time event handle pointer for further sanity checks and cleaning */
+        added_time_event->time_event_handle = ret_time_event_handle;
+    }
+
+    /* exit with out_state value */
+    return out_state;
+
+err_handling:
+    return out_state;
+}
+
+xi_time_event_t* xi_time_event_get_top( xi_vector_t* vector )
+{
+    /* PRE-CONDITIONS */
+    assert( NULL != vector );
+
+    if ( 0 == vector->elem_no )
+    {
+        return NULL;
+    }
+
+    xi_time_event_t* top_one = ( xi_time_event_t* )vector->array[0].selector_t.ptr_value;
+
+    /* the trick is to bubble the element to the end of the vector and then to delete it*/
+    if ( vector->elem_no > 1 )
+    {
+        xi_time_event_move_to_the_end( vector, 0 );
+        xi_vector_del( vector, vector->elem_no - 1 );
+    }
+    else
+    {
+        xi_vector_del( vector, vector->elem_no - 1 );
+    }
+
+    xi_time_event_dispose_time_event( top_one );
+
+    return top_one;
+}
+
+xi_time_event_t* xi_time_event_peek_top( xi_vector_t* vector )
+{
+    /* PRE-CONDITIONS */
+    assert( NULL != vector );
+
+    if ( 0 == vector->elem_no )
+    {
+        return NULL;
+    }
+
+    xi_time_event_t* top_one = ( xi_time_event_t* )vector->array[0].selector_t.ptr_value;
+
+    return top_one;
+}
+
+xi_state_t xi_time_event_restart( xi_vector_t* vector,
+                                  xi_time_event_handle_t* time_event_handle,
+                                  xi_time_t new_time )
+{
+    /* PRE-CONDITIONS */
+    assert( NULL != vector );
+    assert( NULL != time_event_handle );
+
+    /* the element can be found with O(1) complexity cause we've been updating each
+     * element's position during every operation that could've broken it */
+
+    xi_vector_index_type_t index = *time_event_handle->position;
+
+    /* check for the correctness of the time_event_handle */
+    if ( index >= vector->elem_no || index < 0 )
+    {
+        return XI_ELEMENT_NOT_FOUND;
+    }
+
+    /* let's update the key of this element */
+    xi_time_event_t* time_event =
+        ( xi_time_event_t* )vector->array[index].selector_t.ptr_value;
+
+    /* sanity check on the time handle */
+    assert( time_event->time_event_handle == time_event_handle );
+
+    time_event->time_of_execution = new_time;
+
+    xi_time_event_bubble_and_sort_down(
```

Comment:
are these both here because we don't know the relation between the new_time and the old_time? Should they be nested into each other? If not then please separate those it makes the reader brain melt down.